File: app/src/main/java/org/wikiedufoundation/wikiedudashboard/course_detail/common/view/CourseDetailActivity.java
package org.wikiedufoundation.wikiedudashboard.course_detail.common.view;

import android.content.Context;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.Toolbar;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.Toast;

import org.wikiedufoundation.wikiedudashboard.R;
import org.wikiedufoundation.wikiedudashboard.campaign.view.CampaignListFragment;
import org.wikiedufoundation.wikiedudashboard.course_detail.common.data.CourseDetail;
import org.wikiedufoundation.wikiedudashboard.course_detail.common.presenter.CourseDetailPresenter;
import org.wikiedufoundation.wikiedudashboard.course_detail.common.presenter.CourseDetailPresenterImpl;
import org.wikiedufoundation.wikiedudashboard.course_detail.common.provider.RetrofitCourseDetailProvider;
import org.wikiedufoundation.wikiedudashboard.course_detail.common.view.home.CourseHomeFragment;
import org.wikiedufoundation.wikiedudashboard.course_detail.common.view.timeline.CourseTimelineFragment;
import org.wikiedufoundation.wikiedudashboard.course_detail.coures_students.view.StudentListFragment;
import org.wikiedufoundation.wikiedudashboard.course_detail.uploads.view.CourseUploadsFragment;
import org.wikiedufoundation.wikiedudashboard.helper.SharedPrefs;
import org.wikiedufoundation.wikiedudashboard.helper.ViewPagerAdapter;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class CourseDetailActivity extends AppCompatActivity implements CourseDetailView {

    private CourseDetailPresenter courseDetailPresenter;
    private String url;

    @BindView(R.id.toolbar)
    Toolbar toolbar;

    @BindView(R.id.tabLayout)
    TabLayout tabLayout;

    @BindView(R.id.viewPager)
    ViewPager viewPager;

    @BindView(R.id.progressBar)
    ProgressBar progressBar;

    private Context context;
    private SharedPrefs sharedPrefs;
    private ViewPagerAdapter viewPagerAdapter;
    private CourseHomeFragment courseHomeFragment;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_course_detail);
        url = getIntent().getStringExtra("url");
        ButterKnife.bind(this);
        context = this;
        sharedPrefs = new SharedPrefs(context);
        viewPagerAdapter = new ViewPagerAdapter(getSupportFragmentManager());
        viewPager.setAdapter(viewPagerAdapter);
        tabLayout.setupWithViewPager(viewPager);
        toolbar.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                onBackPressed();
            }
        });
        courseDetailPresenter = new CourseDetailPresenterImpl(this, new RetrofitCourseDetailProvider());
        courseDetailPresenter.requestCourseDetail(url);
    }

    @Override
    public void showMessage(String message) {
        Toast.makeText(this, message, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void setData(CourseDetail data) {
        toolbar.setTitle(data.getTitle());
        courseHomeFragment = CourseHomeFragment.newInstance(data);
        setTabs();
    }

    private void setTabs() {
        List<Fragment> fragmentList = new ArrayList<>();
        List<String> titleList = new ArrayList<>();
        titleList.add("Home");
        fragmentList.add(courseHomeFragment);
        if (sharedPrefs.getCookies().equals(sharedPrefs.getWikiEduDashboardCookies())){
            titleList.add("Timeline");
            fragmentList.add(new CourseTimelineFragment());
        }
        titleList.add("Students");
        StudentListFragment studentListFragment=new StudentListFragment();
        Bundle bundle=new Bundle();
        bundle.putString("url",url);
        studentListFragment.setArguments(bundle);
        fragmentList.add(studentListFragment);
        titleList.add("Articles");
        fragmentList.add(new CampaignListFragment());
        titleList.add("Uploads");
        fragmentList.add(CourseUploadsFragment.newInstance(url));
        titleList.add("Activity");
        fragmentList.add(new CampaignListFragment());
        viewPagerAdapter.setTabData(fragmentList, titleList);
        viewPagerAdapter.notifyDataSetChanged();
    }

    @Override
    public void showProgressBar(boolean show) {
        if (show){
            progressBar.setVisibility(View.VISIBLE);
        }else {
            progressBar.setVisibility(View.GONE);
        }
    }
}

File: app/src/main/java/org/wikiedufoundation/wikiedudashboard/ui/coursedetail/common/view/CourseDetailActivity.kt
package org.wikiedufoundation.wikiedudashboard.ui.coursedetail.common.view

import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import kotlinx.android.synthetic.main.activity_course_detail.*
import org.koin.androidx.viewmodel.ext.android.viewModel
import org.wikiedufoundation.wikiedudashboard.R
import org.wikiedufoundation.wikiedudashboard.ui.coursedetail.articlesedited.view.CourseArticlesEditedFragment
import org.wikiedufoundation.wikiedudashboard.ui.coursedetail.common.view.home.CourseHomeFragment
import org.wikiedufoundation.wikiedudashboard.ui.coursedetail.common.viewmodel.CourseDetailViewModel
import org.wikiedufoundation.wikiedudashboard.ui.coursedetail.recentactivity.RecentActivityFragment
import org.wikiedufoundation.wikiedudashboard.ui.coursedetail.students.view.StudentListFragment
import org.wikiedufoundation.wikiedudashboard.ui.coursedetail.uploads.view.CourseUploadsFragment
import org.wikiedufoundation.wikiedudashboard.util.ViewPagerAdapter
import org.wikiedufoundation.wikiedudashboard.util.showToast

/**
 * Activity view for course detail
 * ***/
class CourseDetailActivity : AppCompatActivity(){

    private var enrolled = false
    private lateinit var url: String
    private lateinit var courseHomeFragment: CourseHomeFragment
    private val courseDetailViewModel by viewModel<CourseDetailViewModel>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_course_detail)

        url = intent.getStringExtra("url")
        enrolled = intent.getBooleanExtra("enrolled", false)

        val action = intent.action
        val data = intent.dataString
        if (Intent.ACTION_VIEW == action && data != null) {

            url = data.substring(data.lastIndexOf("courses/") + 8)
            val urlExists = url.contains("?enroll")
            if (urlExists) {
                url = url.lastIndexOf("?enroll").let { url.substring(0, it) }
            }
        }
        url?.let { courseDetailViewModel.requestCourseDetail(it) }
        tabLayout.setupWithViewPager(viewPager)
        toolbar.setNavigationOnClickListener { onBackPressed() }
        setData()
        initializeProgressBar()
        initializeToaster()
    }

    /**
     * Set course detail tabs
     * ***/
    private fun setTabs() {
        val studentListFragment = StudentListFragment()
        val bundle = Bundle()
        bundle.putString("url", url)
        studentListFragment.arguments = bundle

        val courseArticlesEditedFragment = CourseArticlesEditedFragment()
        val bundle2 = Bundle()
        bundle2.putString("url", url)
        courseArticlesEditedFragment.arguments = bundle2

        val recentActivityFragment = RecentActivityFragment()
        val bundle3 = Bundle()
        bundle3.putString("url", url)
        recentActivityFragment.arguments = bundle3

        val fragmentList = listOf(courseHomeFragment,
                studentListFragment,
                courseArticlesEditedFragment,
                url.let { CourseUploadsFragment.newInstance(type = 1, courseDetail = it, courseUploads = null) },
                recentActivityFragment)
        val titleList = listOf("Home", "Students", "Article", "Uploads", "Activity")

        viewPager.apply {
            adapter = ViewPagerAdapter(supportFragmentManager, fragmentList, titleList)
        }
    }

    private fun setData() {
        courseDetailViewModel.coursedetail.observe(this, Observer {
            toolbar.title = it.title
            courseHomeFragment = CourseHomeFragment.newInstance(it)
            setTabs()
        })
    }

    private fun initializeProgressBar() {
        courseDetailViewModel.progressbar.observe(this, androidx.lifecycle.Observer {
            progressBar.visibility = if (it) View.VISIBLE else View.GONE
        })
    }

    private fun initializeToaster() {
        courseDetailViewModel.showMsg.observe(this, androidx.lifecycle.Observer {
            val message = it.showMsg
            showToast(message)
        })
    }
}
